Type error due to possibly C

## Prelude
I downloaded this project and ran it in windows 
- I had no alterations to the source code, but;
- I created empty object files under the build directory, as clang would say that the directory would not exist while building the project.

The above error was my first error when running this project in windows, and below is my second, and hopefully last.

## Error

An example of the error is here:
```
error[E0308]: mismatched types
   --> ./src/b.rs:109:28
    |
109 |         if (*clexes)[i] == (*l).token {
    |            ------------    ^^^^^^^^^^ expected `i64`, found `i32`
    |            |
    |            expected because this is `i64`
    |
help: you can convert an `i32` to an `i64`
    |
109 |         if (*clexes)[i] == (*l).token.into() {
    |                                      +++++++

```
The error, which was only mismatched types, existed 30 times between line 109 and 896.

The full error description is provided below:
[type_error.txt](https://github.com/user-attachments/files/20496457/type_error.txt)

### Ideas
Checking the source, I can see that `token` is a `c_long` or just a `long`, which might have been a `i64` in another machine but is a `i32` in this one. `clexes`, however, is stored as `[i64]`, not a `c_long` indirectly. This may be a simple fix of just editing the parameters of one function and a few other lines of code.

Could this be fixed by making `clexes`  into a `c_long`? Or is this a larger problem that will not be fixed, and can you tell me if my deduction is correct if I do want to spend the effort making the fix?

--- a/src/b.rs
+++ b/src/b.rs
@@ -104,7 +104,7 @@ pub unsafe fn lexer_loc(l: *const stb_lexer, input_path: *const c_char) -> Loc {
     }
 }
 
-pub unsafe fn expect_clexes(l: *const stb_lexer, input_path: *const c_char, clexes: *const [i64]) -> Option<()> {
+pub unsafe fn expect_clexes(l: *const stb_lexer, input_path: *const c_char, clexes: *const [c_long]) -> Option<()> {
     for i in 0..clexes.len() {
         if (*clexes)[i] == (*l).token {
             return Some(());
@@ -130,7 +130,7 @@ pub unsafe fn expect_clexes(l: *const stb_lexer, input_path: *const c_char, clex
     None
 }
 
-pub unsafe fn expect_clex(l: *const stb_lexer, input_path: *const c_char, clex: i64) -> Option<()> {
+pub unsafe fn expect_clex(l: *const stb_lexer, input_path: *const c_char, clex: c_long) -> Option<()> {
     expect_clexes(l, input_path, &[clex])
 }
 
@@ -282,35 +282,35 @@ impl Binop {
     // It's kinda confusing but I don't know how to make it "prettier"
     pub fn from_assign_token(token: c_long) -> Option<Option<Self>> {
         match token {
-            token if token == '=' as i64 => Some(None),
-            CLEX_shleq                   => Some(Some(Binop::BitShl)),
-            CLEX_oreq                    => Some(Some(Binop::BitOr)),
-            CLEX_andeq                   => Some(Some(Binop::BitAnd)),
-            CLEX_pluseq                  => Some(Some(Binop::Plus)),
-            CLEX_minuseq                 => Some(Some(Binop::Minus)),
-            CLEX_muleq                   => Some(Some(Binop::Mult)),
-            CLEX_diveq                   => Some(Some(Binop::Div)),
+            token if token == '=' as c_long => Some(None),
+            CLEX_shleq                      => Some(Some(Binop::BitShl)),
+            CLEX_oreq                       => Some(Some(Binop::BitOr)),
+            CLEX_andeq                      => Some(Some(Binop::BitAnd)),
+            CLEX_pluseq                     => Some(Some(Binop::Plus)),
+            CLEX_minuseq                    => Some(Some(Binop::Minus)),
+            CLEX_muleq                      => Some(Some(Binop::Mult)),
+            CLEX_diveq                      => Some(Some(Binop::Div)),
             _ => None,
         }
     }
 
     pub fn from_token(token: c_long) -> Option<Self> {
         match token {
-            token if token == '+' as i64 => Some(Binop::Plus),
-            token if token == '-' as i64 => Some(Binop::Minus),
-            token if token == '*' as i64 => Some(Binop::Mult),
-            token if token == '/' as i64 => Some(Binop::Div),
-            token if token == '%' as i64 => Some(Binop::Mod),
-            token if token == '<' as i64 => Some(Binop::Less),
-            token if token == '>' as i64 => Some(Binop::Greater),
-            CLEX_greatereq               => Some(Binop::GreaterEqual),
-            CLEX_lesseq                  => Some(Binop::LessEqual),
-            token if token == '|' as i64 => Some(Binop::BitOr),
-            token if token == '&' as i64 => Some(Binop::BitAnd),
-            CLEX_shl                     => Some(Binop::BitShl),
-            CLEX_shr                     => Some(Binop::BitShr),
-            CLEX_eq                      => Some(Binop::Equal),
-            CLEX_noteq                   => Some(Binop::NotEqual),
+            token if token == '+' as c_long => Some(Binop::Plus),
+            token if token == '-' as c_long => Some(Binop::Minus),
+            token if token == '*' as c_long => Some(Binop::Mult),
+            token if token == '/' as c_long => Some(Binop::Div),
+            token if token == '%' as c_long => Some(Binop::Mod),
+            token if token == '<' as c_long => Some(Binop::Less),
+            token if token == '>' as c_long => Some(Binop::Greater),
+            CLEX_greatereq                  => Some(Binop::GreaterEqual),
+            CLEX_lesseq                     => Some(Binop::LessEqual),
+            token if token == '|' as c_long => Some(Binop::BitOr),
+            token if token == '&' as c_long => Some(Binop::BitAnd),
+            CLEX_shl                        => Some(Binop::BitShl),
+            CLEX_shr                        => Some(Binop::BitShr),
+            CLEX_eq                         => Some(Binop::Equal),
+            CLEX_noteq                      => Some(Binop::NotEqual),
             _ => None,
         }
     }
@@ -381,24 +381,24 @@ pub unsafe fn allocate_auto_var(t: *mut AutoVarsAtor) -> usize {
 pub unsafe fn compile_primary_expression(l: *mut stb_lexer, input_path: *const c_char, c: *mut Compiler) -> Option<(Arg, bool)> {
     stb_c_lexer_get_token(l);
     let arg = match (*l).token {
-        token if token == '(' as i64 => {
+        token if token == '(' as c_long => {
             let result = compile_expression(l, input_path, c)?;
-            get_and_expect_clex(l, input_path, ')' as i64)?;
+            get_and_expect_clex(l, input_path, ')' as c_long)?;
             Some(result)
         }
-        token if token == '!' as i64 => {
+        token if token == '!' as c_long => {
             let (arg, _) = compile_primary_expression(l, input_path, c)?;
             let result = allocate_auto_var(&mut (*c).auto_vars_ator);
             push_opcode(Op::UnaryNot{result, arg}, lexer_loc(l, input_path), c);
             Some((Arg::AutoVar(result), false))
         }
-        token if token == '*' as i64 => {
+        token if token == '*' as c_long => {
             let (arg, _) = compile_primary_expression(l, input_path, c)?;
             let index = allocate_auto_var(&mut (*c).auto_vars_ator);
             push_opcode(Op::AutoAssign {index, arg}, lexer_loc(l, input_path), c);
             Some((Arg::Ref(index), true))
         }
-        token if token == '-' as i64 => {
+        token if token == '-' as c_long => {
             let (arg, _) = compile_primary_expression(l, input_path, c)?;
             let index = allocate_auto_var(&mut (*c).auto_vars_ator);
             push_opcode(Op::Negate {result: index, arg}, lexer_loc(l, input_path), c);
@@ -442,7 +442,7 @@ pub unsafe fn compile_primary_expression(l: *mut stb_lexer, input_path: *const c
             let saved_point = (*l).parse_point;
             stb_c_lexer_get_token(l);
 
-            if (*l).token == '(' as i64 {
+            if (*l).token == '(' as c_long {
                 Some((compile_function_call(l, input_path, c, name, name_loc)?, false))
             } else {
                 (*l).parse_point = saved_point;
@@ -476,9 +476,9 @@ pub unsafe fn compile_primary_expression(l: *mut stb_lexer, input_path: *const c
     let saved_point = (*l).parse_point;
     stb_c_lexer_get_token(l);
 
-    if (*l).token == '[' as i64 {
+    if (*l).token == '[' as c_long {
         let (offset, _) = compile_expression(l, input_path, c)?;
-        get_and_expect_clex(l, input_path, ']' as i64)?;
+        get_and_expect_clex(l, input_path, ']' as c_long)?;
 
         let result = allocate_auto_var(&mut (*c).auto_vars_ator);
         push_opcode(Op::Binop {binop: Binop::Plus, index: result, lhs: arg, rhs: offset}, lexer_loc(l, input_path), c);
@@ -616,7 +616,7 @@ pub unsafe fn compile_expression(l: *mut stb_lexer, input_path: *const c_char, c
     let saved_point = (*l).parse_point;
     stb_c_lexer_get_token(l);
 
-    if (*l).token == '?' as i64 {
+    if (*l).token == '?' as c_long {
         let result = allocate_auto_var(&mut (*c).auto_vars_ator);
 
         let addr_condition = (*c).func_body.count;
@@ -629,7 +629,7 @@ pub unsafe fn compile_expression(l: *mut stb_lexer, input_path: *const c_char, c
         push_opcode(Op::Jmp{addr: 0}, lexer_loc(l, input_path), c);
 
         let addr_false = (*c).func_body.count;
-        get_and_expect_clex(l, input_path, ':' as i64)?;
+        get_and_expect_clex(l, input_path, ':' as c_long)?;
 
         let (if_false, _) = compile_expression(l, input_path, c)?;
         push_opcode(Op::AutoAssign {index: result, arg: if_false}, lexer_loc(l, input_path), c);

--- a/src/stb_c_lexer.rs
+++ b/src/stb_c_lexer.rs
@@ -1,5 +1,5 @@
 #![allow(non_upper_case_globals)]
-use core::ffi::{c_char, c_long, c_double, c_int};
+use core::ffi::{c_char, c_long, c_longlong, c_double, c_int};
 
 #[macro_export]
 macro_rules! non_soy_enum {
@@ -61,7 +61,7 @@ pub struct stb_lexer {
     // lexer token variables
     pub token: c_long,
     pub real_number: c_double,
-    pub int_number: c_long,
+    pub int_number: c_longlong,
     pub string: *mut c_char,
     pub string_len: c_int,
 }

--- a/thirdparty/stb_c_lexer.h
+++ b/thirdparty/stb_c_lexer.h
@@ -2,6 +2,7 @@
 // Custom changes:
 // - Fix char literal parsing by Wonshtrum - https://github.com/nothings/stb/pull/1653
 // - Update stb_c_lexer_get_location to operate on input_stream instead of stb_lexer
+// - Update `int_number` field of the `stb_lexer` struct to be `long long` instead of `long`
 
 // stb_c_lexer.h - v0.12+ - public domain Sean Barrett 2013
 // lexer for making little C-like languages with recursive-descent parsers
@@ -131,7 +132,7 @@ typedef struct
    // lexer token variables
    long token;
    double real_number;
-   long   int_number;
+   long long int_number;
    char *string;
    int string_len;
 } stb_lexer;
@@ -227,7 +228,7 @@ typedef double     stb__clex_int;
 #define intfield   real_number
 #define STB__clex_int_as_double
 #else
-typedef long       stb__clex_int;
+typedef long long  stb__clex_int;
 #define intfield   int_number
 #endif
 
@@ -709,7 +710,7 @@ int stb_c_lexer_get_token(stb_lexer *lexer)
 
                   #ifdef STB__clex_hex_ints
                   #ifdef STB__CLEX_use_stdlib
-                  lexer->int_number = strtol((char *) p, (char **) &q, 16);
+                  lexer->int_number = strtoll((char *) p, (char **) &q, 16);
                   #else
                   {
                      stb__clex_int n=0;
@@ -763,7 +764,7 @@ int stb_c_lexer_get_token(stb_lexer *lexer)
          if (p[0] == '0') {
             char *q = p;
             #ifdef STB__CLEX_use_stdlib
-            lexer->int_number = strtol((char *) p, (char **) &q, 8);
+            lexer->int_number = strtoll((char *) p, (char **) &q, 8);
             #else
             stb__clex_int n=0;
             while (q != lexer->eof) {
@@ -785,7 +786,7 @@ int stb_c_lexer_get_token(stb_lexer *lexer)
          {
             char *q = p;
             #ifdef STB__CLEX_use_stdlib
-            lexer->int_number = strtol((char *) p, (char **) &q, 10);
+            lexer->int_number = strtoll((char *) p, (char **) &q, 10);
             #else
             stb__clex_int n=0;
             while (q != lexer->eof) {
@@ -842,7 +843,7 @@ static void print_token(stb_lexer *lexer)
       #if defined(STB__clex_int_as_double) && !defined(STB__CLEX_use_stdlib)
       case CLEX_intlit    : printf("#%g", lexer->real_number); break;
       #else
-      case CLEX_intlit    : printf("#%ld", lexer->int_number); break;
+      case CLEX_intlit    : printf("#%lld", lexer->int_number); break;
       #endif
       case CLEX_floatlit  : printf("%g", lexer->real_number); break;
       default:

